pkg/service/structs.go

```diff
@@ -0,0 +1,188 @@
+package service
+
+import (
+	"fmt"
+
+	"capnproto.org/go/capnp/v3"
+	"github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+	api "github.com/wetware/ww/internal/api/service"
+)
+
+type Location struct {
+	api.SignedLocation
+}
+
+func NewLocation() (Location, error) {
+	_, seg := capnp.NewSingleSegmentMessage(nil)
+	sloc, err := api.NewSignedLocation(seg)
+	if err != nil {
+		return Location{}, fmt.Errorf("failed to create location: %w", err)
+	}
+
+	_, seg = capnp.NewSingleSegmentMessage(nil)
+	loc, err := api.NewLocation(seg)
+	if err != nil {
+		return Location{}, fmt.Errorf("failed to create location: %w", err)
+	}
+
+	if err := sloc.SetLocation(loc); err != nil {
+		return Location{}, fmt.Errorf("failed to set location: %w", err)
+	}
+
+	return Location{SignedLocation: sloc}, nil
+}
+
+func (loc Location) Validate(topic string) error {
+	capLoc, err := loc.Location()
+	if err != nil {
+		return fmt.Errorf("failed to read location: %w", err)
+	}
+	service, err := capLoc.Service()
+	if err != nil {
+		return fmt.Errorf("failed to read service name: %w", err)
+	}
+	if topic != service {
+		return fmt.Errorf("the topic and the service name are different, topic: %s - service: %s", topic, service)
+	}
+
+	if ok, err := loc.VerifySinature(); err != nil {
+		return fmt.Errorf("failed to verify signature: %w", err)
+	} else if !ok {
+		return fmt.Errorf("invalid signature", err)
+	}
+
+	return nil
+}
+
+func (loc Location) Sign(pk crypto.PrivKey) error {
+	capLoc, err := loc.Location()
+	if err != nil {
+		return fmt.Errorf("failed to read location: %w", err)
+	}
+
+	b, err := capLoc.Message().Marshal()
+	if err != nil {
+		return fmt.Errorf("failed to marshal location: %w", err)
+	}
+
+	signature, err := pk.Sign(b)
+	if err != nil {
+		return fmt.Errorf("failed to sign location: %w", err)
+	}
+
+	if err := loc.SetSignature(signature); err != nil {
+		return fmt.Errorf("failed to set signature: %w", err)
+	}
+
+	return nil
+}
+
+func (loc Location) VerifySinature() (bool, error) {
```

Comment:
I think we can just return `error`, and assume that validation passes if the error is nil.